torch/fx/passes/reinplace.py

```diff
@@ -0,0 +1,540 @@
+    nodes_used_after = set()
+    for t in tensor_aliases:
+        # get all nodes that use the current alias
+        usage_nodes = t.meta['node_usages']
+        for n in usage_nodes:
+            # We only care about usages after the current node
+            if n.meta['node_idx'] <= op_index:
+                continue
+            if n in tensor_aliases:
+                continue
+            if only_include_view_inverse_nodes and not _satisfies_inverse_view(n, tensor_aliases):
+                continue
+            nodes_used_after.add(n)
+    return nodes_used_after
+
+
+
+@compatibility(is_backward_compatible=True)
+def reinplace(gm, *sample_args):
+    """
+    Given an fx.GraphModule, modifies it to perform "reinplacing",
+    mutating the nodes of the graph.
+    We look for out-of-place op call sites like `b = a.add(...)`,
+    and convert them to be inplace (`b = a.add_(...)`),
+    as long as the input to the current operator ("a") isn't re-used
+    anywhere later in the graph.
+
+    Sample inputs are needed to determine aliasing relationships of the inputs.
+    In general, we can't reinplace node `b = a.add(...)` if "a" aliases any of the
+    inputs to the program.
+
+    Given a node "b = foo(a, ...)", the algorithm for re-inplacing is as follows:
+
+    (1) Check if foo has a mutating variant. If not, move to the next node.
+
+        Note that we ignore view ops (we don't bother to turn `as_strided()`
+        into `as_strided_()`), as it complicates the algorithm and doesn't
+        provide meaningful speedups.
+
+        Currently, we also only check for an inplace op, `foo_`.
+        Later, we should beef this up to check for out= or mutable ops.
+
+    (2) Check if "a" is an alias of any of the program inputs.
+
+        If it is, skip and move to the next node.
+        Inplace'ing an op that would cause it to mutate a program is not sound,
+        because that would be a side effect visible to the user.
+
+        NOTE: there's a future optimization that we should make:
+        if "a" is a (alias of a)  program input, but later in the program
+        there is a node that looks like "a.copy_(...)",
+        Then re-inplacing is ok to do - we are temporarily re-using a's buffer,
+        which will later be overwritten by the copy_() call.
+
+        This will be an important optimization to have for programs that mutate
+        their inputs. It currently isn't implemented though.
+
+    (3) Check that "a" and all of its outstanding aliases are not used anywhere
+        later in the graph. If this is the case, then it's safe to re-inplace
+        to "b = foo_(a)".
+
+        There are a few caveats to this, explained in more detail below:
+        (a) If "a" is used later as an argument to a view op, that is okay.
+            It's only a problem if "a" (or that view) is later passed
+            into a normal operator, or if it is returned as the program output.
```

Comment:
If the view is never passed into a normal operator, doesn't that mean it's dead? So you can simplify condition (a) if you DCE before reinplacing.

Reply:
Yep that's true - what's a little annoying is that you might have a chain of a view op + a scatter op that both use "a" later in the graph, like:
```
b = a.add(...)
a_view = a.view(...)
a_select_scatter = select_scatter(a_view, ....)
```

In that case, both the `a_view` and `a_select_scatter` nodes show up as "nodes that use an alias of a later in the graph". The logic I have currently will recognize the `a_select_scatter` node as being ok to remove, but not the `a_view` node. So I figured it'd be simpler just to ignore any view ops when figuring out which nodes can prevent an op from being reinplaced.

I guess another option is that when we detect that the `a_select_scatter` node can be removed, we include it + it's entire view chain.